File: src/EditorFeatures/CSharpTest/IntroduceVariable/IntroduceLocalForExpressionTests.cs
﻿// Licensed to the .NET Foundation under one or more agreements.
// The .NET Foundation licenses this file to you under the MIT license.
// See the LICENSE file in the project root for more information.

#nullable disable

using System.Threading.Tasks;
using Microsoft.CodeAnalysis.CodeRefactorings;
using Microsoft.CodeAnalysis.CodeStyle;
using Microsoft.CodeAnalysis.CSharp.CodeStyle;
using Microsoft.CodeAnalysis.CSharp.IntroduceVariable;
using Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.CodeRefactorings;
using Microsoft.CodeAnalysis.Editor.UnitTests.CodeActions;
using Microsoft.CodeAnalysis.Test.Utilities;
using Roslyn.Test.Utilities;
using Xunit;

namespace Microsoft.CodeAnalysis.Editor.CSharp.UnitTests.IntroduceVariable
{
    public partial class IntroduceLocalForExpressionTests : AbstractCSharpCodeActionTest
    {
        protected override CodeRefactoringProvider CreateCodeRefactoringProvider(Workspace workspace, TestParameters parameters)
            => new CSharpIntroduceLocalForExpressionCodeRefactoringProvider();

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task IntroduceLocal_NoSemicolon()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        new DateTime()[||]
    }
}",
@"
using System;

class C
{
    void M()
    {
        DateTime {|Rename:dateTime|} = new DateTime();
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task IntroduceLocal_NoSemicolon_BlankLineAfter()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        new DateTime()[||]

    }
}",
@"
using System;

class C
{
    void M()
    {
        DateTime {|Rename:dateTime|} = new DateTime();

    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task IntroduceLocal_NoSemicolon_SelectExpression()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        [|new DateTime()|]
    }
}",
@"
using System;

class C
{
    void M()
    {
        DateTime {|Rename:dateTime|} = new DateTime();
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        [WorkItem(35525, "https://github.com/dotnet/roslyn/issues/35525")]
        public async Task IntroduceLocal_Inside_Expression()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        new TimeSpan() +[||] new TimeSpan();
    }
}",
@"
using System;

class C
{
    void M()
    {
        TimeSpan {|Rename:timeSpan|} = new TimeSpan() + new TimeSpan();
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task IntroduceLocal_Semicolon()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        new DateTime();[||]
    }
}",
@"
using System;

class C
{
    void M()
    {
        DateTime {|Rename:dateTime|} = new DateTime();
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task IntroduceLocal_Semicolon_BlankLineAfter()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        new DateTime();[||]

    }
}",
@"
using System;

class C
{
    void M()
    {
        DateTime {|Rename:dateTime|} = new DateTime();

    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task IntroduceLocal_Semicolon_SelectExpression()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        [|new DateTime()|];
    }
}",
@"
using System;

class C
{
    void M()
    {
        DateTime {|Rename:dateTime|} = new DateTime();
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task IntroduceLocal_Semicolon_SelectStatement()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        [|new DateTime();|]
    }
}",
@"
using System;

class C
{
    void M()
    {
        DateTime {|Rename:dateTime|} = new DateTime();
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task MissingOnAssignmentExpressionStatement()
        {
            await TestMissingInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        int a = 42;
        [||]a = 42;
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task IntroduceLocal_Space()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        new DateTime(); [||]
    }
}",
@"
using System;

class C
{
    void M()
    {
        DateTime {|Rename:dateTime|} = new DateTime(); 
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task IntroduceLocal_LeadingTrivia()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        // Comment
        new DateTime();[||]
    }
}",
@"
using System;

class C
{
    void M()
    {
        // Comment
        DateTime {|Rename:dateTime|} = new DateTime();
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task IntroduceLocal_PreferVar()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        new DateTime();[||]
    }
}",
@"
using System;

class C
{
    void M()
    {
        var {|Rename:dateTime|} = new DateTime();
    }
}", options: new OptionsCollection(GetLanguage())
    {
        { CSharpCodeStyleOptions.VarElsewhere, CodeStyleOptions2.TrueWithSuggestionEnforcement },
        { CSharpCodeStyleOptions.VarWhenTypeIsApparent, CodeStyleOptions2.TrueWithSuggestionEnforcement },
    });
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task MissingOnVoidCall()
        {
            await TestMissingInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        Console.WriteLine();[||]
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task MissingOnDeclaration()
        {
            await TestMissingInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        var v = new DateTime()[||]
    }
}");
        }

        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)]
        public async Task IntroduceLocal_ArithmeticExpression()
        {
            await TestInRegularAndScriptAsync(
@"
using System;

class C
{
    void M()
    {
        1 + 1[||]
    }
}",
@"
using System;

class C
{
    void M()
    {
        int {|Rename:v|} = 1 + 1;
    }
}");
        }
    }
}

File: src/Features/VisualBasicTest/IntroduceVariable/IntroduceLocalForExpressionTests.vb
' Licensed to the .NET Foundation under one or more agreements.
' The .NET Foundation licenses this file to you under the MIT license.
' See the LICENSE file in the project root for more information.

Imports Microsoft.CodeAnalysis.CodeRefactorings
Imports Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.CodeRefactorings
Imports Microsoft.CodeAnalysis.VisualBasic.IntroduceVariable

Namespace Microsoft.CodeAnalysis.Editor.VisualBasic.UnitTests.IntroduceVariable
    <Trait(Traits.Feature, Traits.Features.CodeActionsIntroduceLocalForExpression)>
    Partial Public Class IntroduceLocalForExpressionTests
        Inherits AbstractVisualBasicCodeActionTest

        Protected Overrides Function CreateCodeRefactoringProvider(workspace As Workspace, parameters As TestParameters) As CodeRefactoringProvider
            Return New VisualBasicIntroduceLocalForExpressionCodeRefactoringProvider()
        End Function

        <Fact>
        Public Async Function IntroduceLocal() As Task
            Await TestInRegularAndScriptAsync(
"
imports System

class C
    sub M()
        N()[||]
    end sub

    function N() as string
    end function
end class",
"
imports System

class C
    sub M()
        Dim {|Rename:v|} = N()
    end sub

    function N() as string
    end function
end class")
        End Function

        <Fact>
        Public Async Function IntroduceLocal_TrailingBlankLine() As Task
            Await TestInRegularAndScriptAsync(
"
imports System

class C
    sub M()
        N()[||]

    end sub

    function N() as string
    end function
end class",
"
imports System

class C
    sub M()
        Dim {|Rename:v|} = N()

    end sub

    function N() as string
    end function
end class")
        End Function

        <Fact>
        Public Async Function IntroduceLocal_Selection() As Task
            Await TestInRegularAndScriptAsync(
"
imports System

class C
    sub M()
        [|N()|]
    end sub

    function N() as string
    end function
end class",
"
imports System

class C
    sub M()
        Dim {|Rename:v|} = N()
    end sub

    function N() as string
    end function
end class")
        End Function

        <Fact>
        Public Async Function IntroduceLocal_Space() As Task
            Await TestInRegularAndScriptAsync(
"
imports System

class C
    sub M()
        N() [||]
    end sub

    function N() as string
    end function
end class",
"
imports System

class C
    sub M()
        Dim {|Rename:v|} = N() 
    end sub

    function N() as string
    end function
end class")
        End Function

        <Fact>
        Public Async Function IntroduceLocal_LeadingTrivia() As Task
            Await TestInRegularAndScriptAsync(
"
imports System

class C
    sub M()
        ' Comment
        N()[||]
    end sub

    function N() as string
    end function
end class",
"
imports System

class C
    sub M()
        ' Comment
        Dim {|Rename:v|} = N()
    end sub

    function N() as string
    end function
end class")
        End Function

        <Fact>
        Public Async Function MissingOnVoidCall() As Task
            Await TestMissingInRegularAndScriptAsync(
"
imports System

class C
    sub M()
        Console.WriteLine()[||]
    end sub
end class")
        End Function

        <Fact>
        Public Async Function MissingOnDeclaration() As Task
            Await TestMissingInRegularAndScriptAsync(
"
imports System

class C
    sub M()
        dim x = N()[||]
    end sub

    function N() as string
    end function
end class")
        End Function
    End Class
End Namespace
